### DISCUSSION TITLE
findOne/findOneOrFail should work without where param or with empty where param

### DISCUSSION THREAD

**Discussion by ersinakinci** [mikro-orm/mikro-orm#5607] [Ideas]
**Is your feature request related to a problem? Please describe.**

In my unit tests, I commonly set up some stuff like a single user for my web app. I know for certain that there is only one record in a given table. I often don't want or need to write logic for picking the exact record based on where statements, I just want to retrieve the first (and only) record in the most straightforward manner possible.

To achieve this goal, it would be nice if I could trivially run `em.findOneOrFail(User)` or even `em.findOneOrFail(User, {})` in my tests, but I get the `You cannot call 'EntityManager.findOne()' with empty 'where' parameter` error.

I'm not sure why this error exists. Is it to prevent users from mistakenly constructing a query that doesn't have a where parameter?

`em.findOne`/`em.findOneOrFail` without a where parameter intuitively makes sense to me. It means "find me one _____, whichever one comes first," similar to JS iterables' `find` method.

If preventing users from shooting themselves in the foot is the main goal, forcing them to pass the where parameter but allowing an empty object--as in, `em.findOne(User, {})`--would be a good compromise.

**Describe the solution you'd like**

The ability to use `findOne` and `findOneOrFail` without a where parameter or with an empty object for a where parameter. The behavior would be defined as ```select `t0`.* from `table` as `t0` limit = 1```.

**Describe alternatives you've considered**

I know that I can use `find` to get around this limitation, but it's cumbersome to use `find`, pull the first element from the array, and test that it's not `undefined`.

**Comment by ersinakinci** [comment]
I'll add one more thing, the current behavior is inconsistent with the TS typings. The method signature allows you to pass an empty where param object; `em.findOne(User, {})` is valid TS. The error only comes up at runtime.

If the intention is to prevent empty `findOne`/`findOneOrFail` queries, then this issue is actually less of a feature request and more of a bug report because the types should ensure that the where param object is populated.

That said, I would prefer that the requirement is loosened to allow for empty where param objects or perhaps no where param at all.

**Reply by B4nan** [reply]
> I'll add one more thing, the current behavior is inconsistent with the TS typings. The method signature allows you to pass an empty where param object; em.findOne(User, {}) is valid TS. The error only comes up at runtime.

AFAIK there is no way to enforce non-empty object in TS. So if you consider that a bug, I am calling it a wontfix :]

**Comment by B4nan** [comment]
An empty query is quite problematic, it can lead to security issues since most of the time you don't want to fetch some random user, I can understand that only inside tests - that's the main reason behind this. I am not interested in changing this behavior, I would even consider it breaking, but we can have an opt-in flag to allow it, PR welcome for that.

**Comment by B4nan** [comment]
> If preventing users from shooting themselves in the foot is the main goal, forcing them to pass the where parameter but allowing an empty object--as in, em.findOne(User, {})--would be a good compromise.

This itself is also problematic, imagine you would do this:

```ts
em.findOne(User, { email, password });
```

And you (for whatever reason) end up with both of `email` and `password` as `undefined`. You surely don't want to return a random user from your database in that case. But `{ email: undefined, password: undefined }` is literally the same as `{}`. Nowdays people can opt-into this behavior via `ignoreUndefinedInQuery` option.

edit: here is a typeorm issue about this problem to give you better understanding of the reason behind this: https://github.com/typeorm/typeorm/issues/9316

**Comment by giancarlo-dm** [comment]
@B4nan what if I have a `configs` table that will always have a single record? Right now I did the following workaround: 
```typescript
const myConfig = await em.findOneOrFail(ConfigTable, { id: { $exists: true  }});
```

While it works, the ORM could supply a method like `em.findAny` and `em.findAnyOrFail`. This should be enough to prevent mistakes done by dev and still grant some security as it is explicit that I want _any_ record.

Of course one could argue why I have a table with a single record? Well, legacy db and not enough time available for proper design. In another project I work we have a `configs`table with a `type` column to identify config for what I need and it the config itself is stored as JSONB, so there I always need to supply a conditions object.

**Reply by B4nan** [reply]
You can always do this:

```ts
const [myConfig] = await em.findAll(ConfigTable);
```

(maybe combined with `limit: 1` to be sure you wont overfetch in case things change in your database)

**Reply by giancarlo-dm** [reply]
Yes, but it wouldn't throw the `fail exception` and would require me to manually throw the exception by writing an `if`. So for the moment I am sticking with my approach.

**Comment by Dennis4720** [comment]
+1

**Reply by B4nan** [reply]
You can enable this nowadays with `ignoreUndefinedInQuery` in your orm config.
